civicrm_entity.module

```diff
@@ -179,3 +179,66 @@ function _civicrm_entity_get_entity_type_from_elements($elements) {
 
   return NULL;
 }
+
+/**
+ * Implements hook_civicrm_pre().
+ */
+function civicrm_entity_civicrm_pre($op, $objectName, $id, &$params) {
+  $operations = ['create', 'update', 'delete'];
+  if (!in_array($op, $operations) || !$entityType = SupportedEntities::getDrupalType($objectName)) {
+    return;
+  }
+
+  /** @var \Drupal\civicrm_entity\CiviEntityStorage $storage */
+  $storage = \Drupal::entityTypeManager()->getStorage($entityType);
+  $entity = $storage->create($params);
+  if ($id) {
```

Comment:
Delete operations don't provide parameters, so we have to at least set an ID. Maybe a better solution would be to load the full Drupal entity for `edit` and `delete` operations.

Reply:
Loading the full Drupal entity for 'edit' and 'delete' makes sense, because the Drupal hooks are probably expecting that anyway. In fact, I'd be a little worried that not doing that could create weird interactions with certain hook implementations down the line. For example, some hook expecting the entity to be in a particular state after being loaded, but it isn't, because the entity wasn't actually loaded.

Reply:
I would prefer to load the Drupal version of the entity for sure....If you notice in the 7.x version, in civicrm_entity_civicrm_post(), that is where we trigger Rules events, and you'll notice we did NOT load the object there....this is because of problems with multi-valued custom fields....The entity_metadata_wrapper in D7 would load a wrapper from the $objectRef, but it's not ideal, and has caused a few problems in the past..